### ISSUE TITLE
How exactly are items encrypted?

### ISSUE THREAD

**Issue by matze** [standardnotes/web#117]
I am currently writing a client in Python but even though the documentation claims that a re-implementation should be idiot-proof, I am not capable of doing it at least for encryption scheme `001`. Authentication works and I can get the encrypted data items. First I extract the encryption and authentication keys from the `data` dictionary:
```python
enc_item_key = base64.b64decode(data['enc_item_key'])
# enc_item_key contains 144 bytes data
iv = enc_item_key[:16]
aes = AES.new(unhexlify(master_key), AES.MODE_CBC, iv)
item_key = aes.decrypt(enc_item_key)
# item_key contains 144 bytes hex encoded data result
item_key = item_key[16:-ord(item_key[-1])]
# item_key contains 112 bytes hex encoded data remain (without IV and padding)
```

as you can see from the comments I encounter the first problem of not having enough key data! Surprisingly, I can verify that data is not tempered with by providing an auth key that extends into the encryption key:

```python
auth_key = item_key[len(item_key) / 2 - 8:]
hmac_sha256 = hmac.new(unhexlify(auth_key), data['content'], hashlib.sha256)

if hmac_sha256.hexdigest() != data['auth_hash']:
    raise AuthenticationError("Could not verify authentication hash")
```

However, providing a 32 byte key by extending into the auth key does not work and I only decrypt garbage:
```python
enc_key = item_key[:len(item_key) / 2 + 8]
content = base64.b64decode(data['content'][3:])
iv = content[:16]
aes = AES.new(unhexlify(enc_key), AES.MODE_CBC, iv)
d = aes.decrypt(content)
```

So what's going on here?

**Comment by mobitar** [comment]
Are you following the instructions found here? http://standardfile.org/#item-encryption

I don't see why you're splitting the string according to byte length. I would instead split it by the colon character. Are you decrypting 001 or 002 encryption type?

**Comment by matze** [comment]
> Are you following the instructions found here? http://standardfile.org/#item-encryption

I did until I found out that most of my items are encrypted according to 001.

**Comment by mobitar** [comment]
So then are you following this? https://github.com/standardfile/standardfile.github.io/blob/master/doc/spec-001.md#item-encryption

**Comment by matze** [comment]
Yes and as you can read from my initial issue, I am able to successfully verify the data with the auth key. But here's the thing:  the `enc_item_key` is just 144 bytes long, after decryption only 112 bytes of hex-encoded data are left, which makes both keys only 28 bytes long. To remedy this I shift the auth key data until I have the required 32 bytes.

**Comment by mobitar** [comment]
In 001, IVs are not used at all. The reason being is that each item is encrypted with a random key every time it changes, effectively serving as an IV. This was controversial, so 002 makes better use of IVs.  So this line:

```
iv = enc_item_key[:16]
```

should be:

```
iv = ""
```

**Comment by matze** [comment]
I know that fact and it effectively makes no difference if I pass `" "*16` or the "original" IV. And as I said, it decrypts the key material for the auth key just fine but it's ... not enough.

**Comment by mobitar** [comment]
I'm not sure I'm following.

Here's a hex encoded key:

```
"84bd8ce40a74186da470deb8440e09513c7ba4cefc318a37c86bf2c5b6285904b026e5a7784e5cd3a58c3f3e9baba5985ce026fc36e71ad4b7dd8e5f9d6c3547".bytesize
```

outputs `128` in Ruby. Convert the hex to binary:

```
s.to_i(16).to_s(2).bytesize
```

outputs 512.

**Comment by matze** [comment]
Okay, let's assume this is the content of an `enc_item_key`:
```
YAdXfBzfFPuinCQWfv/ZsRHsMDQbjJwBQM/PU0LtvQtBnVTQDVYIUlKIw6W79OS7PLR2AcDdW2OZrurqx0G85y7knL4AqMw2m6eZCWlH1vM153PzfOTAn5dKzt3RXalDjC8Ga93hQ0AIJ11zuEfFvVLqRbkvQEtbTs+Waae393k074nbaRPlVz/nWbdebKFp
```
Let's decode the base64 encoding:
```python
import base64
d = base64.b64decode(open('payload').read())
len(d)
```
which prints 144 bytes. If I do 
```python
aes = AES.new(unhexlify(master_key), AES.MODE_CBC, ' '*16)
item_key = aes.decrypt(d)
len(item_key)
```
it prints 144 bytes as well. Of those, the first 16 bytes are the useless IV and the last 16 bytes padding. Printing the last 16 bytes
```python
print ' '.join('{}'.format(ord(x)) for x in item_key[-16:])
```
returns
```
16 16 16 16 16 16 16 16 16 16 16 16 16 16 16 16
```
So out of the 144 bytes I have only 112 bytes left which really *do* contain hex-encoded key material.

**Comment by mobitar** [comment]
There is no auth key contained. I can't fully focus on following your python implementation right now, but I can forward you to how other repos do it:

iOS:
https://github.com/standardnotes/iOS/blob/master/standardnotes/Crypto.swift#L73

Web/Desktop:
https://github.com/standardnotes/web/blob/master/app/assets/javascripts/app/services/helpers/crypto.js#L33

**Comment by matze** [comment]
> There is no auth key contained.

Then what does the decrypted item key contain? According to your documentation we have to "set encryption key ek = first half and authentication key ak = second half".

**Comment by matze** [comment]
So, [here](https://github.com/matze/python-standardfile) is what I have so far but the problem is not solved yet. Is there a way to enforce re-encryption according to 002?

**Comment by matze** [comment]
Besides PyCrypto, I tried the cryptography and M2Crypto libraries all with the same outcome, so I suspect it's not the AES implementation I use but something else probably related to CryptoJS. Stupid question but are you using a "Word Array" or a string to pass the key to the CryptoJS encrypt function? The latter causes a non-standard salt and IV to be [generated](https://stackoverflow.com/a/35536933/997768) …

**Comment by mobitar** [comment]
> According to your documentation we have to "set encryption key ek = first half and authentication key ak = second half".

That's for 002. For 001, the auth_hash is a separate parameter that comes back from the server with each item. The enc_item_key is only used to decrypt the content and nothing else.

**Comment by mobitar** [comment]
Everything is passed in as byte data and not literal strings. We've gotten this working on Android, iOS, and Web, so it's not an issue with CryptoJS. Have you looked at the other implementations?

**Comment by mobitar** [comment]
```
    enc_item_key = base64.b64decode(data['enc_item_key'])
    iv = enc_item_key[:16]
    aes = AES.new(unhexlify(master_key), AES.MODE_CBC, iv)
    item_key = aes.decrypt(enc_item_key)        # 144 bytes hex encoded data result
    item_key = item_key[16:-ord(item_key[-1])]  # 112 bytes hex encoded data remain (without IV and padding)

    item_key = '\0'*8 + unhexlify(item_key)
    enc_key = item_key[8:len(item_key) / 2 + 8]
    auth_key = item_key[len(item_key) / 2:]
```

I have no idea what you're doing here. Don't know why you're doing manual addition/division. Again you're passing in a wrong IV to AES. I don't know what this does: ` item_key[16:-ord(item_key[-1])]`.

**Comment by matze** [comment]
> Don't know why you're doing manual addition/division.

Again, because I want to extract the `auth_key` (which in fact is the hash value but you named it authentication key in the docs …) I have to compare to and the `enc_key` to decrypt the content of the Item.

> Again you're passing in a wrong IV to AES.

Again, it *does not matter* as I wrote previously. I can pass anything as IV and I will return the correct output sans the new IV but I have to pass something that is 16 bytes long (which is by the way required per the CBC mode).

> Don't know why you're doing manual addition/division.

To split the decrypted in half just as you are writing in the docs. I add more bytes because – and I wrote this multiple times now – I have only 112 bytes of hex-encoded key data instead of 128.

> I don't know what this does: `item_key[16:-ord(item_key[-1])]`

Again – as written in the previous in the very first comment – I am removing the first 16 bytes which is the IV that is generated by the PyCrypto (and all other implementations to be exact) decryption routine as well as the remaining padding. `item_key[-1]` is the very last element, `ord()` turns it into a number – the number of pad bytes –, thus I remove as many elements from the back as given.

Yes, I did have a look at your other examples however, they are relatively convoluted, spread out across multiple modules and use their own AES implementations.
